### ISSUE TITLE
False-positive trait-impl-incorrect-safety error for Send & Sync

### ISSUE THREAD

**Issue by K0bin** [rust-lang/rust-analyzer#19499]
Rust-analyzer complaints:

> unsafe impl for safe trait

when implementing either `Send` or `Sync`. Both of those are unsafe traits, so it's a false positive. The compiler is fine with it too. If I remove the `unsafe`, Rust-Analyzer complaints that it is required, like you'd expect ("the trait `Sync` requires an `unsafe impl` declaration").

**rust-analyzer version**: 0.4.2363-standalone (8a84c2bc28 2025-03-31)

I switched the VS Code extension to the pre-release channel. It does happen with the latest stable version too. The problem only started a few days ago.

**rustc version**: rustc 1.85.1 (4eb161250 2025-03-15)

**editor or extension**: Codium 1.98.2 c6286962b0f5f2c3ad5e558a439dc3fcfd20b2ac

**relevant settings**: No value for `CARGO`, `RUSTC` or `RUSTUP_HOME`.

**repository link (if public, optional)**: https://github.com/K0bin/SourceRenderer/blob/master/graphics/vulkan/src/buffer.rs#L31

Link points to one line where it happens.

**code snippet to reproduce**: 
```rust
struct Test {
    test: *const u8
}

unsafe impl Send for Test {}
unsafe impl Sync for Test {}


fn main() {
    println!("Hello, world!");
}
```

**Comment by ChayimFriedman2** [comment]
Works fine for me. Sounds like you have a corrupted sysroot - can you try executing `rustup component remove rust-src` then `rustup component install rust-src`?

**Comment by K0bin** [comment]
Fwiw I also updated the Rust compiler with `rustup update` recently and that didn't make any difference.

**Comment by K0bin** [comment]
Re-installing rust-src went fine:

```
[robin@robinpc unsafe-send-test]$ rustup component remove rust-src
info: removing component 'rust-src'
[robin@robinpc unsafe-send-test]$ rustup component add rust-src
info: downloading component 'rust-src'
info: installing component 'rust-src'
[robin@robinpc unsafe-send-test]$ 
```

Rust-Analyzer can't load it though.

```
2025-04-01T16:00:43.761783928+02:00 ERROR can't load standard library, try installing `rust-src` sysroot_path=/usr
```

I did restart the Rust-Analyzer server after reinstalling rust-src.

**Comment by ChayimFriedman2** [comment]
What is your toolchain version?

**Comment by lnicola** [comment]
> `sysroot_path=/usr`

That doesn't sound right, are you mixing `rustup` and distro-built `rustc`?

**Comment by K0bin** [comment]
> That doesn't sound right, are you mixing rustup and distro-built rustc?

```
[robin@robinpc SourceRenderer]$ which rustc
/home/robin/.cargo/bin/rustc
[robin@robinpc SourceRenderer]$ which rustup
/home/robin/.cargo/bin/rustup
```

> What is your toolchain version?

``stable-x86_64-unknown-linux-gnu (active, default)``

I don't know how to get the exact version. I ran `rustup update` today so I assume its the latest stable one?

**Comment by ChayimFriedman2** [comment]
> > What is your toolchain version?
> 
> `stable-x86_64-unknown-linux-gnu (active, default)`
> 
> I don't know how to get the exact version. I ran `rustup update` today so I assume its the latest stable one?

Yes, it is. But maybe what @lnicola said is the clue.

**Comment by lnicola** [comment]
Does it make a difference if you start Codium from the terminal? `rustup` sets up `PATH` in your shell, but this won't apply to your DE.

**Comment by K0bin** [comment]
> Does it make a difference if you start Codium from the terminal? rustup sets up PATH in your shell, but this won't apply to your DE.

Not sure I understand correctly. As far as I understand it it should normally be Codium (or rather the Rust plugin of it) that starts Rust Analyzer. Why would it make a difference if I start Codium using the terminal rather than the dock (no notable PATH changes in the .desktop file).

That said. I can no longer reproduce the issue. Unfortunately I have no idea what fixed it. Clearly seems to have been some kind of weird setup issue on my system. Sorry for the noise.

**Comment by lnicola** [comment]
> As far as I understand it it should normally be Codium (or rather the Rust plugin of it) that starts Rust Analyzer.

It is.

> Why would it make a difference if I start Codium using the terminal rather than the dock (no notable PATH changes in the .desktop file).

Exactly, the `.desktop` file has no notable `PATH` additions, but your shell does, since `rustup` changes it to add `~/.cargo/bin` to your `PATH`. So if you run Codium from the launcher or dock, it won't see your shell's `PATH`. The solution is to add it to `/etc/environment` or `~/.config/environment.d`:

```bash
$ cat ~/.config/environment.d/10-path.conf 
PATH=$PATH:$HOME/.cargo/bin:$HOME/.local/bin
```

**Comment by maia-s** [comment]
I also get this issue. It's not consistent, usually it works, but sometimes rust-analyzer seems to get into a state where this happens frequently. When it happens, restarting rust-analyzer through cmd-shift-P fixes it for a while, but when it has started happening it always happens again soon, until I restart vscode.
